README.md
# NAME

Wasm::Wasm3 - Self-contained [WebAssembly](https://webassembly.org/) via [wasm3](https://github.com/wasm3/wasm3)

# SYNOPSIS

Basic setup:

    my $env = Wasm::Wasm3->new();
    my $module = $env->parse_module($wasm_binary);
    my $runtime = $env->create_runtime(1024)->load_module($module);

Run [WASI](https://wasi.dev):

    my $exit_code = $runtime->run_wasi('arg1', 'arg2');

WebAssembly-exported globals:

    my $global = $module->get_global('some-value');

    $module->set_global('some-value', 1234);

WebAssembly-exported memory:

    $runtime->set_memory( $offset, $bytes );

    my $from_wasm = $runtime->get_memory( $offset, $length );

Call a WebAssembly-exported function:

    my @out = $runtime->call('some-func', @args);

Implement a WebAssembly-imported function in Perl:

    $runtime->link_function('mod-name', 'func-name', 'v(ii)', $coderef);

(`v(ii)` is the function’s signature; see [Wasm::Wasm3::Runtime](https://metacpan.org/pod/Wasm%3A%3AWasm3%3A%3ARuntime) for
details.)

# DESCRIPTION

Well-known WebAssembly runtimes like [Wasmer](https://wasmer.io),
[Wasmtime](https://wasmtime.dev), or [WAVM](https://github.com/wavm/wavm)
often require nonstandard dependencies/toolchains (e.g., LLVM or Rust).
Their builds can take a while, especially on slow machines, and only
the most popular platforms may enjoy support.

[wasm3](https://github.com/wasm3/wasm3) takes a different tactic from
the aforementioned “big dogs”: whereas those are all JIT compilers,
wasm3 is a WebAssembly _interpreter_. This makes it quite small and
fast/simple to build, which lets you run WebAssembly in environments
that something bigger may not support. Runtime performance lags the
“big dogs” significantly, but startup latency will likely be lower, and
memory usage is **much** lower.

This distribution includes wasm3, so you don’t need to build it yourself.

# STATUS

This Perl library is EXPERIMENTAL.

Additionally, wasm3 is, as of this writing, rather less complete than
Wasmer et al. wasm3 only exports a single WebAssembly memory, for
example. It can’t import memories or globals, and it neither imports
_nor_ exports tables.

# [WASI](https://wasi.dev) SUPPORT

wasm3 implements WASI via either of (as of this writing) two backends:
a wrapper around [uvwasi](https://github.com/nodejs/uvwasi), and a
less-complete original implementation. The former needs
[libuv](https://libuv.org), which doesn’t compile on all platforms,
while the latter should compile everywhere this module can run.

This distribution’s `Makefile.PL` implements logic to determine which
backend to use.

You’re free, of course, to implement your own WASI imports rather than to
use wasm3’s. Depending on how much of WASI you actually need that may not
be as onerous as it sounds; see the distribution’s `t/faux_wasi.t` for an
example.

# MEMORY LEAK DETECTION

To help you avoid memory leaks, instances of all classes `warn()`
if their `DESTROY()` method runs at global destruction time.

This necessitates extra care when linking Perl functions to WASM;
see [Wasm::Wasm3::Module](https://metacpan.org/pod/Wasm%3A%3AWasm3%3A%3AModule) for details, and the distribution’s
`t/faux_wasi.t` for an example.

# DOCUMENTATION

This module generally documents only those aspects of its usage that
are germane to this module specifically. For more details, see
wasm3’s documentation.

# STATIC FUNCTIONS & CONSTANTS

## ($MAJOR, $MINOR, $REV) = M3\_VERSION

Returns wasm3’s version as 3 integers.

## $STRING = M3\_VERSION\_STRING

Returns wasm3’s version as a string.

## `TYPE_I32`, `TYPE_I64`, `TYPE_F32`, `TYPE_F64`

Numeric constants that indicate the corresponding WebAssembly type.

## $YN = WASI\_BACKEND

Either `uvwasi` or `simple`. See above about WASI support for
details.

# METHODS

## $OBJ = _CLASS_->new()

Instantiates _CLASS_.
Creates a new wasm3 environment and binds it to the returned object.

## $RUNTIME = _OBJ_->create\_runtime( $STACKSIZE )

Creates a new wasm3 runtime from _OBJ_.
Returns a [Wasm::Wasm3::Runtime](https://metacpan.org/pod/Wasm%3A%3AWasm3%3A%3ARuntime) instance.

## $MODULE = _OBJ_->parse\_module( $WASM\_BINARY )

Loads a WebAssembly module from _binary_ (`*.wasm`) format.
Returns a [Wasm::Wasm3::Module](https://metacpan.org/pod/Wasm%3A%3AWasm3%3A%3AModule) instance.

If your WebAssembly module is in text format rather than binary,
you’ll need to convert it first. Try
[wabt](https://github.com/webassembly/wabt) if you need such a tool.

# LICENSE & COPYRIGHT

Copyright 2022 Gasper Software Consulting. All rights reserved.

This library is licensed under the same terms as Perl itself.
See [perlartistic](https://metacpan.org/pod/perlartistic).


Changes
Revision history for Perl extension Wasm::Wasm3:

0.03
- Update wasm3 from upstream.

0.02    Sat, 18 Apr 2022
- BUG FIX: Big-endian global i32 imports now work (via wasm3 PR #322).
- Add WASI support (via wasm3 PR #331).
- Add Cygwin compatibility (via wasm3 PR #330).
- Add Strawberry Perl compatibility.
- Fix compatibility with 5.16.3 multiplicity perls.
- Update to latest wasm3.
- Switch (temporarily?) to forked wasm3.

0.01    Mon, 11 Apr 2022
- Initial release


t/assets/wasi-demo.ts
import "wasi";

import { Console, Descriptor, FileSystem } from "as-wasi/assembly";

console.log("Hello, world!");

console.log("argv=" + process.argv.join(","));

process.env.keys().forEach( function(k) {
    console.log("env:" + k + "=" + process.env.get(k));
} );

const dirname = "/épée";

const contents = FileSystem.readdir(dirname);
if (contents) {
    console.log(`readdir(${dirname}): ` + (contents as string[]).sort().join());
}
else {
    console.log(`readdir(${dirname}) failed`);
}

const from_stdin = Descriptor.Stdin.readString();
if (from_stdin === null) {
    Console.log("got nothing on stdin!");
}
else {
    const str = from_stdin as string;
    Console.error(`from stdin: ${str}`);
}

process.exitCode = 42;


Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker::CPANfile;
use ExtUtils::CBuilder;
use File::Temp;
use File::Spec;

my $uvwasi_inc = File::Spec->catdir( qw( wasm3 _deps uvwasi-src include ) );

my %OS_CAN_UVWASI = map { $_ => 1 } (
    'linux',
    'darwin',
    'freebsd',
    'openbsd',
);

my $use_uvwasi = $OS_CAN_UVWASI{$^O} || 0;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Wasm::Wasm3',
    VERSION_FROM      => 'lib/Wasm/Wasm3.pm', # finds $VERSION
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Wasm/Wasm3.pm', # retrieve abstract from module
       AUTHOR         => [
            'Felipe Gasper (FELIPE)',
        ],
      ) : ()
    ),
    LICENSE => "perl_5",

    INC => join(
        q< >,
        ( $use_uvwasi ? "-I$uvwasi_inc" : () ),
        '-Wall',
        '-std=c99',
    ),

    DEFINE => join(
        q< >,
        "-DWW3_UVWASI=$use_uvwasi",
        ($use_uvwasi ? '-Dd_m3HasUVWASI' : ()),
    ),

    LIBS => ($use_uvwasi ? '-lpthread' : q<>),

    OBJECT => join(
        q< >,
        '$(BASEEXT)$(OBJ_EXT)',
        File::Spec->catfile( qw(wasm3 source libm3.a) ),
        (
            $use_uvwasi
            ? (
                File::Spec->catfile( qw(wasm3 _deps uvwasi-build libuvwasi_a.a) ),
                File::Spec->catfile( qw(wasm3 _deps libuv-build libuv_a.a) ),
            )
            : (),
        ),
    ),

    PMLIBDIRS => ['lib'],

    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url => 'git://github.com/FGasper/p5-Wasm-Wasm3.git',
                web => 'https://github.com/FGasper/p5-Wasm-Wasm3',
            },
            bugtracker => {
                web => 'https://github.com/FGasper/p5-Wasm-Wasm3/issues',
            },
        },
    },
);

# ----------------------------------------------------------------------

package MY;

use Alien::cmake3;
use Config;

sub postamble {
    my $cmake_path = File::Spec->catfile(
        Alien::cmake3->bin_dir(),
        Alien::cmake3->exe(),
    );

    my @cflags = (
        '-fPIC',
        ($^O eq 'cygwin' ? '-D__MINGW32__' : ()),
    );

    my @cmake_opts = (
        ( $use_uvwasi ? () : (-D => 'BUILD_WASI=simple' ) ),

        # _GNU_SOURCE is for Cygwin to have ptsname(), which doesn’t really
        # matter for now since Cygwin can’t do WASI, but we might as well:
        -D => qq/CMAKE_C_FLAGS='@cflags'/,

        -D => "APP_DIR='../wasm3-dummy-app-dir'",
    );

    if ($^O eq 'MSWin32') {
        push @cmake_opts, -G => q<'MinGW Makefiles'>,
    }

    return <<"MAKE_FRAG"

.PHONY:
build-wasm3:
\t$cmake_path @cmake_opts -S wasm3 -B wasm3
\t\$(MAKE) -C wasm3
MAKE_FRAG
}

my $orig_top_targets;
BEGIN {
$orig_top_targets = __PACKAGE__->can('top_targets');
}

sub top_targets {
    my $out = $orig_top_targets->(@_);

    $out =~ s<^(all\s*:+s*)><$1 build-wasm3 >;

    return $out;
}


lib/Wasm/Wasm3/Module.pm
package Wasm::Wasm3::Module;

use strict;
use warnings;

=encoding utf-8

=head1 NAME

Wasm::Wasm3::Module

=head1 SYNOPSIS

See L<Wasm::Wasm3>.

=head1 DESCRIPTION

This module exposes L<wasm3|https://github.com/wasm3/wasm3>’s
module object to Perl.

=cut

#----------------------------------------------------------------------

use Wasm::Wasm3;

#----------------------------------------------------------------------

=head1 METHODS

This class is not directly instantiated; see L<Wasm::Wasm3> for
details.

=head2 $value = I<OBJ>->get_global( $NAME )

Returns the value of the $NAMEd export global.

=head2 $type = I<OBJ>->get_global_type( $NAME )

Returns the type (e.g., Wasm::Wasm3::TYPE_I32) of the $NAMEd export global.

=head2 $obj = I<OBJ>->link_function( $MODULE_NAME, $FUNCTION_NAME, $SIGNATURE, $CODEREF )

Sets $CODEREF as $MODULE_NAME.$FUNCTION_NAME’s implementation inside the
WebAssembly module. See below for L</$SIGNATURE>.

$CODEREF will I<always> be called in list context. $CODEREF B<MUST> return
the number of arguments that $SIGNATURE indicates, or you’ll get an error
(possibly an unhelpful one).

If $CODEREF throws, the exception is C<warn()>ed, and a generic
callback-failed error is thrown to the C<link_function()> caller.

=head3 WASM Context in Callbacks

Your callback may need to reference either the wasm3 runtime or module.
When doing this, be sure to use a C<weaken()>ed copy (cf. L<Scalar::Util>)
of that object, or you’ll leak memory (and eventually get a C<warn()>ing
about it).

For example:

    my $weak_runtime = $runtime;
    Scalar::Util::weaken($weak_runtime);

    $module->link_function(
        mymodule => myfuncname => 'v(ii)',
        sub {
            my ($buf_p, $buflen) = @_;

            my $buf = $weak_runtime->get_memory($buf_p, $buflen);

            # Now do something cool with $buf.

            return;
        },
    );

The distribution’s F<t/faux_wasi.t> shows this technique in action.

(An alternative design would be to pass a special context object
to every callback, but the weak-reference approach is more efficient.)

=head3 $SIGNATURE

$SIGNATURE is wasm3’s own convention to describe a function’s inputs &
outputs. As of this writing wasm3’s documentation doesn’t describe it very
well, so we’ll describe it here.

The format is C<$RETURNS($ARGS)>, where $RETURNS and $ARGS are both either:

=over

=item * C<v>, to indicate empty (C<v> meaning “void”)

=item * … a sequence of one or more of: C<i> (i32), C<I> (i64), C<f> (f32),
C<F>, (f64)

=back

Space characters are ignored.

For example: C<v(if)> indicates a function that takes i32 and f32 as
arguments and returns nothing.

=head2 $obj = I<OBJ>->link_wasi_default()

A quick helper to link L<WASI|https://wasi.dev> includes via
wasm3’s L<uvwasi|https://github.com/nodejs/uvwasi> integration.

This uses wasm3’s built-in WASI defaults, e.g., STDIN becomes WASI file
descriptor 0.

=head2 $obj = I<OBJ>->link_wasi( %OPTS )

(NB: Only available if uvwasi is your WASI backend; see L<Wasm::Wasm3>
for details.)

Like C<link_wasi_default()> but takes a list of key/value pairs that
offer the following controls:

%OPTS are:

=over

=item * C<in>, C<out>, C<err> - File handles to the WASI input, output,
and error streams. Defaults are file descriptors 0, 1, and 2 respectively.

=item * C<env> - A reference to an array of key/value byte-string pairs
to pass as the WASI environment.

=item * C<preopen> - A reference to a hash of WASI paths to system/real
paths.

B<IMPORTANT:> WASI paths are character strings, while system paths are
B<byte> strings. The discrepancy arises because WASI paths are always
character strings, while Perl treats all system paths as byte strings
(even on OSes like Windows where paths are character strings).

So if, for example, you have directory F</tmp/føø> that you’ll access
in WASI as F</épée>, your code might look thus:

    preopen => {
        do { use utf8; '/épée' } => do { no utf8; '/tmp/føø' },
    },

=back

=cut

our $WASI_MODULE_STR;

sub link_wasi_default {
    my ($self) = @_;

    return $self->_perl_link_wasi('_link_wasi_default');
}

sub link_wasi {
    my ($self, @args) = @_;

    return $self->_perl_link_wasi('_link_wasi', @args);
}

sub _perl_link_wasi {
    my ($self, $fn, @args) = @_;

    if ($WASI_MODULE_STR) {
        if ($WASI_MODULE_STR ne "$self") {
            die "$self: WASI is already linked! ($WASI_MODULE_STR)";
        }
    }
    else {
        $self->$fn(@args);
        $WASI_MODULE_STR = "$self";
    }

    return $self;
}

sub DESTROY {
    my ($self) = @_;

    $self->_destroy_xs();

    if ($WASI_MODULE_STR && ($WASI_MODULE_STR eq "$self")) {
        undef $WASI_MODULE_STR;
    }

    return;
}

1;


lib/Wasm/Wasm3/Runtime.pm
package Wasm::Wasm3::Runtime;

use strict;
use warnings;

=encoding utf-8

=head1 NAME

Wasm::Wasm3::Runtime

=head1 SYNOPSIS

See L<Wasm::Wasm3>.

=head1 DESCRIPTION

This module exposes L<wasm3|https://github.com/wasm3/wasm3>’s
runtime object to Perl.

=cut

#----------------------------------------------------------------------

use Wasm::Wasm3;

#----------------------------------------------------------------------

=head1 METHODS

This class is not directly instantiated; see L<Wasm::Wasm3> for
details.

=head2 $obj = I<OBJ>->load_module( $MODULE_OBJ )

Loads a parsed module (i.e., L<Wasm::Wasm3::Module> instance).
Returns I<OBJ>.

=head2 @returns = I<OBJ>->call( $FUNCTION_NAME, @ARGUMENTS )

Calls the named function with the given arguments, returning the
returns from that function.

=head2 @types = I<OBJ>->get_function_arguments( $FUNCTION_NAME )

Returns the named function’s argument types, as TYPE_* constants.
(cf. L<Wasm::Wasm3>)

=head2 @types = I<OBJ>->get_function_returns( $FUNCTION_NAME )

Like C<get_function_arguments()> but for return types.

=head2 $str = I<OBJ>->get_memory( [ $OFFSET [, $WANTLENGTH] ] )

Fetches all or part of I<OBJ>’s WebAssembly memory buffer as a byte string.
$OFFSET defaults to 0, and $WANTLENGTH defaults to the buffer’s length less
$OFFSET. If $WANTLENGTH + $OFFSET exceed the buffer’s size, the returned
string will contain just the content from $OFFSET to the buffer’s end.

Currently both values B<MUST> be nonnegative.

=head2 $count = I<OBJ>->get_memory_size()

Returns the size, in bytes of I<OBJ>’s WebAssembly memory buffer.

=head2 $obj = I<OBJ>->set_memory( $OFFSET, $NEW_BYTES )

Overwrites all or part of I<OBJ>’s WebAssembly memory buffer with
$NEW_BYTES. Returns I<OBJ>.

=cut

1;


lib/Wasm/Wasm3.pm
package Wasm::Wasm3;

use strict;
use warnings;

=encoding utf-8

=head1 NAME

Wasm::Wasm3 - Self-contained L<WebAssembly|https://webassembly.org/> via L<wasm3|https://github.com/wasm3/wasm3>

=head1 SYNOPSIS

Basic setup:

    my $env = Wasm::Wasm3->new();
    my $module = $env->parse_module($wasm_binary);
    my $runtime = $env->create_runtime(1024)->load_module($module);

Run L<WASI|https://wasi.dev>:

    my $exit_code = $runtime->run_wasi('arg1', 'arg2');

WebAssembly-exported globals:

    my $global = $module->get_global('some-value');

    $module->set_global('some-value', 1234);

WebAssembly-exported memory:

    $runtime->set_memory( $offset, $bytes );

    my $from_wasm = $runtime->get_memory( $offset, $length );

Call a WebAssembly-exported function:

    my @out = $runtime->call('some-func', @args);

Implement a WebAssembly-imported function in Perl:

    $runtime->link_function('mod-name', 'func-name', 'v(ii)', $coderef);

(C<v(ii)> is the function’s signature; see L<Wasm::Wasm3::Runtime> for
details.)

=head1 DESCRIPTION

Well-known WebAssembly runtimes like L<Wasmer|https://wasmer.io>,
L<Wasmtime|https://wasmtime.dev>, or L<WAVM|https://github.com/wavm/wavm>
often require nonstandard dependencies/toolchains (e.g., LLVM or Rust).
Their builds can take a while, especially on slow machines, and only
the most popular platforms may enjoy support.

L<wasm3|https://github.com/wasm3/wasm3> takes a different tactic from
the aforementioned “big dogs”: whereas those are all JIT compilers,
wasm3 is a WebAssembly I<interpreter>. This makes it quite small and
fast/simple to build, which lets you run WebAssembly in environments
that something bigger may not support. Runtime performance lags the
“big dogs” significantly, but startup latency will likely be lower, and
memory usage is B<much> lower.

This distribution includes wasm3, so you don’t need to build it yourself.

=head1 STATUS

This Perl library is EXPERIMENTAL.

Additionally, wasm3 is, as of this writing, rather less complete than
Wasmer et al. wasm3 only exports a single WebAssembly memory, for
example. It can’t import memories or globals, and it neither imports
I<nor> exports tables.

=head1 L<WASI|https://wasi.dev> SUPPORT

wasm3 implements WASI via either of (as of this writing) two backends:
a wrapper around L<uvwasi|https://github.com/nodejs/uvwasi>, and a
less-complete original implementation. The former needs
L<libuv|https://libuv.org>, which doesn’t compile on all platforms,
while the latter should compile everywhere this module can run.

This distribution’s F<Makefile.PL> implements logic to determine which
backend to use.

You’re free, of course, to implement your own WASI imports rather than to
use wasm3’s. Depending on how much of WASI you actually need that may not
be as onerous as it sounds; see the distribution’s F<t/wasi_pp.t> for an
example.

=head1 MEMORY LEAK DETECTION

To help you avoid memory leaks, instances of all classes C<warn()>
if their C<DESTROY()> method runs at global destruction time.

This necessitates extra care when linking Perl functions to WASM;
see L<Wasm::Wasm3::Module> for details, and the distribution’s
F<t/wasi_pp.t> for an example.

=head1 DOCUMENTATION

This module generally documents only those aspects of its usage that
are germane to this module specifically. For more details, see
wasm3’s documentation.

=cut

#----------------------------------------------------------------------

use XSLoader;

our $VERSION;

BEGIN {
    $VERSION = '0.01_92';

    XSLoader::load( __PACKAGE__, $VERSION );
}

use Wasm::Wasm3::Module ();
use Wasm::Wasm3::Runtime ();

use constant M3_VERSION => (_M3_VERSION_MAJOR, _M3_VERSION_MINOR, _M3_VERSION_REV);

#----------------------------------------------------------------------

=head1 STATIC FUNCTIONS & CONSTANTS

=head2 ($MAJOR, $MINOR, $REV) = M3_VERSION

Returns wasm3’s version as 3 integers.

=head2 $STRING = M3_VERSION_STRING

Returns wasm3’s version as a string.

=head2 C<TYPE_I32>, C<TYPE_I64>, C<TYPE_F32>, C<TYPE_F64>

Numeric constants that indicate the corresponding WebAssembly type.

=head2 $YN = WASI_BACKEND

Either C<uvwasi> or C<simple>. See above about WASI support for
details.

=head1 METHODS

=head2 $OBJ = I<CLASS>->new()

Instantiates I<CLASS>.
Creates a new wasm3 environment and binds it to the returned object.

=head2 $RUNTIME = I<OBJ>->create_runtime( $STACKSIZE )

Creates a new wasm3 runtime from I<OBJ>.
Returns a L<Wasm::Wasm3::Runtime> instance.

=head2 $MODULE = I<OBJ>->parse_module( $WASM_BINARY )

Loads a WebAssembly module from I<binary> (F<*.wasm>) format.
Returns a L<Wasm::Wasm3::Module> instance.

If your WebAssembly module is in text format rather than binary,
you’ll need to convert it first. Try
L<wabt|https://github.com/webassembly/wabt> if you need such a tool.

=cut

=head1 LICENSE & COPYRIGHT

Copyright 2022 Gasper Software Consulting. All rights reserved.

This library is licensed under the same terms as Perl itself.
See L<perlartistic>.

=cut

1;
